daemon/k8s_watcher.go

```diff
@@ -1735,127 +1708,109 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
+			cnpError = policyImportErr
 		}
 
+		updatedCNP, cnpUpdateErr = updateCNPNodeStatus(cnp, overallErr == nil, policyImportErr == nil, cnpError, rev)
 		if cnpUpdateErr == nil {
-			logger.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
-			break
+			logger.WithField("status", updatedCNP.Status).Debug("successfully updated with status")
+			return overallErr
 		}
 
 		// Wait a small amount of time to try to update again.
 		logger.WithError(cnpUpdateErr).Warningf("update of CNP failed; sleeping for %s amount of time before trying again", updateWaitDuration)
 		time.Sleep(updateWaitDuration)
 	}
 
-	if cnpUpdateErr != nil {
-		return cnpUpdateErr
-	} else {
-		return waitForEPsErr
+	// If the process of waiting for endpoints has been successful but the
+	// CNP status report failed, report an error to cause the controller to
+	// retry.
+	if overallErr != nil && cnpUpdateErr != nil {
+		overallErr = cnpUpdateErr
 	}
+
+	return overallErr
 }
 
-func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, err error, rev uint64, annotations map[string]string) error {
-	var (
-		cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
-		err2  error
-	)
+func updateCNPNodeStatus(cnp *cilium_v2.CiliumNetworkPolicy, enforcing, ok bool, importError error, rev uint64) (*cilium_v2.CiliumNetworkPolicy, error) {
+	ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
 
+	// Retrieve the latest revision from the server to minimize the risk of
+	// a locally corrupted CNP makes its way back into the apiserver
+	latestServerRev, err := ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Get(cnp.GetName(), metav1.GetOptions{})
 	if err != nil {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			Enforcing:   enforcing,
-			Error:       err.Error(),
-			OK:          ok,
-			LastUpdated: cilium_v2.NewTimestamp(),
-			Annotations: annotations,
-		}
+		return nil, err
+	}
+
+	// Create a copy so it can be modified to update the status section
+	latestServerCopy := latestServerRev.DeepCopy()
+
+	cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
+		Enforcing:   enforcing,
+		OK:          ok,
+		LastUpdated: cilium_v2.NewTimestamp(),
+		Annotations: latestServerCopy.Annotations,
+	}
+
+	if importError != nil {
+		cnpns.Error = importError.Error()
 	} else {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			Enforcing:   enforcing,
-			Revision:    rev,
-			OK:          ok,
-			LastUpdated: cilium_v2.NewTimestamp(),
-			Annotations: annotations,
-		}
+		cnpns.Revision = rev
 	}
 
-	nodeName := node.GetName()
-	cnp.SetPolicyStatus(nodeName, cnpns)
-	ns := k8sUtils.ExtractNamespace(&cnp.ObjectMeta)
+	latestServerCopy.SetPolicyStatus(node.GetName(), cnpns)
 
 	switch {
 	case ciliumUpdateStatusVerConstr.Check(k8sServerVer):
-		_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).UpdateStatus(cnp)
+		_, err = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).UpdateStatus(cnp)
```

Comment:
You can modify the whole spec of the CNP and it will not matter for this case, the `UpdateStatus` only modifies the status of a k8s object even if the client completely destroys the spec of the object.

Reply:
I saw that, and yes, once we EOL older k8s versions we can potentially simplify and remove this code again.